enhancements/kube-apiserver/stability-point-to-point-network-check.md

```diff
@@ -0,0 +1,305 @@
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+#### RelatedEndpoint Custom Resource
+
+```go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ConditionCheck references a "interesting" endpoint for a given pod.
+type ConditionCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   ConditionCheckSpec   `json:"spec"`
+	Status ConditionCheckStatus `json:"status"`
+}
+
+type ConditionCheckSpec struct {
+	// Selector for pods which should check this condition
+	Selector *metav1.LabelSelector      `json:"selector"`
+	Endpoint ConditionCheckSpecEndpoint `json:"endpoint"`
+}
+
+type ConditionCheckSpecEndpoint struct {
+	// URL is the url of the end point
+	URL string `json:"url"`
+	// CACertFile, if specified, will be used to validate server certs
+	CACertFile string `json:"url,omitempty"`
+	// CertFile, if specified, will be sent to the endpoint as a client cert
+	CertFile string `json:"url,omitempty"`
+	// KeyFile, if specified, will be used when CertFile is also specified
+	KeyFile string `json:"url,omitempty"`
+}
+
+type ConditionCheckStatus struct {
+	Latencies  []ConditionCheckMetric    `json:"latencies"`
+	Conditions []ConditionCheckCondition `json:"conditions"`
+}
+
+type ConditionCheckMetric struct {
+	Value float64     `json:"value"`
+	Time  metav1.Time `json:"time,omitempty"`
+}
+
+type ConditionCheckConditionType string
+
+const (
+	Reachable ConditionCheckConditionType = "Reachable"
+)
+
+type ConditionCheckCondition struct {
+	Type               ConditionCheckConditionType `json:"type"`
+	Status             ConditionStatus             `json:"status"`
+	LastTransitionTime metav1.Time                 `json:"lastTransitionTime,omitempty"`
+	Reason             string                      `json:"reason,omitempty"`
+	Message            string                      `json:"message,omitempty"`
+}
+
+type ConditionStatus string
+
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+```
+
+Example `RelatedEndpoint` CRs instance that an operator might create for one of its managed pods:
+
+```yaml
+Kind: RelatedEndpoint
+Metadata:
+    Name: kas-10-0-137-239-r7-to-etcd-10-0-148-99
+    Namespace: openshift-kube-apiserver
+Spec:
+    Selector:
+        MatchLabels:
+            app : openshift-kube-apiserver
+            revision : 11
```

Comment:
https://github.com/openshift/enhancements/pull/289/files#diff-4cd882c0f5672d93efe9d85e62732945R180

this doesn't line up? is this binary running as container in the kas pod? if yes, what is this selector doing?

Reply:
The pod needs to identify which one of these "check" resources it should execute. typically, pods would have unique names, but in the case of our static pods, the pod name is not enough to identify.